### DISCUSSION TITLE
Nested Loop Join vs Hash Join (토스 면접 기출)

### DISCUSSION THREAD

**Discussion by nak-honest** [woowacourse-6th-computer-sign-pen/comsa-interview-study#13] [Database]
Nested Loop Join vs Hash Join (토스 면접 기출)

**Comment by slimsha2dy** [comment]
참석자 1
질문 먼저 해주세요.

참석자 2
두 차이가 뭔가요?

참석자 1
Nested Loop Join이 뭐냐 하면, 두 테이블이 있으면 한쪽 테이블을 이렇게 루프하면서 한 레코드씩 살펴보면서 그 레코드의 조인 조건에 맞는 레코드들을 오른쪽 테이블에서 찾는 거예요. 그래서 보통 Nested Loop Join, 중첩 루프잖아요. 보통 왼쪽에 먼저 들어가는 루프가 아우터 테이블이 되고, 왜냐하면 아우터 루프니까 오른쪽에 있는 게 이너 테이블이 됩니다. 아우터 테이블을 순회하면서 한 레코드씩 오른쪽 테이블, 이너 테이블을 루프해서 찾는 거죠. 이게 Nested Loop Join이고, Hash Join도 이너 테이블과 아우터 테이블이 있죠. 이때 이너 테이블에 있는 조인 키가 있을 거예요. 먼저 말씀드리면 Hash Join은 동등 조건일 때만 가능합니다. 조인 조건이 동등 조건일 때, 조인 키를 기준으로 이너 테이블에 있는 값들로 해시 테이블을 만들어요.

참석자 1
예를 들어 학생 테이블과 과목 테이블이 있다고 하면, 과목 키로 조인을 해야 돼요. 그러면 과목 키를 갖고 해시 테이블을 하나 만들어요. 그리고 아우터 테이블을 순회하면서 그 조인 키로 해시 테이블을 하나씩 조회하는 겁니다. 아이디를 조회해서 해당 값이 있으면 가져오는 식으로, 이렇게 해시 테이블을 한 번씩 조회해주는 게 Hash Join입니다.

참석자 3
그럼 더 빠른가요?

참석자 1
왜냐하면 Hash 테이블의 시간 복잡도가 O(1)이니까요. 한 번 루프를 하면 끝나겠죠. 반면 Nested Loop Join은 한 번 돌 때마다, 한 레코드에 대해서 조건에 맞는 걸 찾아야 돼요. 오른쪽 테이블에서 루프를 하면서 인덱스가 있으면 더 빠르겠지만, 그래서 그런 차이가 있습니다.

참석자 1
항상 빠르진 않겠죠. 왜냐하면 Hash Join은 해시 테이블을 만들어야 되니까요. 정확한 기준은 모르겠지만 Nested Loop Join은 두 테이블의 크기가 작을 때 아주 효율적이라고 들었고, 해시 테이블을 만드는 오버헤드보다 더 빨리 찾을 수 있으면 Nested Loop Join이 더 빠를 것 같습니다.

참석자 3
그러면 조인할 때 실행 계획을 옵티마이저가 계산하잖아요. 그러면 주가 아우터 테이블이 있고 이너 테이블이 있잖아요. 그걸 어떤 방식으로 하게 될지도 알 수가 있을까요?

참석자 1
그건 제가 잘 모르는 부분이긴 한데, 제가 듣기로는...

참석자 3
음악도 했잖아요.

참석자 1
제가 알기로는 조인 조건에서 왼쪽에 오는 드라이빙 테이블이 아우터 테이블이 되는 걸로 알고 있어요. 레코드 수가 많지 않아요.

참석자 3
맞아요. 근데 왼쪽 순서는 상관없긴 한데, 드라이빙이 기준이 되는 거죠. 옵티마이저가 실행 계획을 수립할 때 드라이빙과 드리븐을 선택해서 성능을 최적화하죠. 드라이빙은 데이터를 가져오겠다고 선택하는 테이블이고, 옵티마이저가 드라이빙 테이블과 드리븐 테이블을 선택할 때 성능을 고려해서 결정합니다. 예를 들어 드리븐 테이블에 PK가 있으면 성능이 좋을 거예요. PK로 바로 접근할 수 있으니까요. 그러면 PK가 있는 테이블이 드리븐이 되거나...

참석자 1
근데 어차피 조인 조건으로 찾는 거 아니에요? 조인 조건이 PK가 아니면 조인 조건이 PK인 값을...

참석자 3
근데 항상 PK로만 하진 않잖아요.

참석자 1
그렇죠. 조인 조건이 PK일 때만 그렇게 된다는 거죠.

참석자 3
네, 보통 그렇게 되거나 레코드 건수가 적은 테이블이 드리븐이 될 때가 있어요. 그 용어가 기억이 안 나네요. 결정 조건이라던가 범위 탐색 같은 게 있었는데...

참석자 2
소트 머지 조인도 아시나요?

참석자 3
맞아요.

참석자 1
그거는 제가 대충 찾아봤는데 잘 안 쓰는 것 같더라고요.

참석자 3
그거는 뭔가요?

참석자 1
Sort Merge Join이에요. 잘 안 쓰는 것 같은데 자세히는 안 봤어요. 네스트 루프 조인의 단점을 보완한 느낌인데, 네스트 루프 조인은 루프를 돌면서 오른쪽을 조회하잖아요. 그런데 조회할 때마다 랜덤 액세스를 해야 해서 디스크 참조가 많아지죠. 그래서 두 테이블을 메모리에 올려놓고 정렬한 다음에 조인을 하는 방식으로 알고 있어요.

참석자 3
본 것 같아요. 하나로 합친다는 건가? 조인 조건을 기준으로 정렬을 해서...

참석자 2
하나로 합친다는...

참석자 3
그러니까 머지 않겠어요? 정렬이 되어 있으니까 병합하는 거죠. 그래서 정렬된 데이터를 병합하는 방식인 것 같아요. 삐트리 설명도 해주세요.

참석자 1
B-트리요. 잘 몰라요. 뜬금없지 않나요? B-트리를 설명할게요. B-트리는 이진 트리에서 출발했는데, 이진 트리를 사용하면 이진 탐색을 빠르게 할 수 있잖아요.

참석자 3
네.

참석자 1
그런데 B-트리는 한 노드에 값을 여러 개 넣는 거죠. 이진 트리는 각 노드에 값이 하나 있고, 그 값보다 작으면 왼쪽, 크면 오른쪽으로 가잖아요. 그런데 B-트리는 한 노드에 여러 개의 값을 넣어놓고, 값들 사이에 간선이 있는 거죠. 값을 찾아서 그 간선이 갈 곳을 찾은 다음에 내려가요. 이진 트리에서 값이 여러 개 들어갈 수 있는 건데, 간선도 여러 개이고 B-트리입니다. 그러면 왜 인덱스에서 B-트리를 쓸까요? B-트리는 이진 트리와 달리 높이가 낮아서 탐색 시간이 짧아요.

참석자 1
높이가 클수록 밑으로 내려가는 작업이 많고, 노드에서 비교하는 연산은 적어지는 거죠. B-트리는 디스크 접근을 최소화하기 위해 설계된 구조예요. 트리 구조에서는 노드 간의 물리적 거리가 멀어 디스크 I/O가 많이 필요할 수 있는데, B-트리는 노드 내에 여러 값을 저장해서 높이를 낮추고 디스크 I/O를 줄이는 방식입니다.

참석자 1
노드에서 노드로 넘어가는 작업은 DB에서는 비용이 크지만, 노드 안에서의 연산은 메모리 상에서 이루어지므로 비용이 적습니다. 높이를 낮추면 디스크 I/O가 줄어들지만, 노드 크기를 키우면 높이를 줄일 수 있어요. 결국 인덱스가 B-트리를 사용하는 이유는 디스크 I/O를 최소화하기 위해서입니다.

참석자 3
좋은데요. 그런데 그게 이유가 되나요? 바이너리 트리랑 B-트리랑 탐색 시간이 균일한가요? 바이너리 트리는 균형을 맞출 수 없나요?

참석자 1
바이너리 트리도 균형을 맞출 수 있어요. 보통 일반 이진 트리와 비교하지 않고, 균형 잡힌 트리인 AVL 트리나 레드-블랙 트리와 비교하죠. B-트리도 균형을 유지하니까요. 두 트리 모두 평균적으로 탐색 시간이 비슷하지만, B-트리는 디스크 접근을 최소화하는 데 최적화되어 있습니다.

참석자 1
회크 조인은 언제 쓰나요? 회크 조인은 테이블 크기가 크고 동등 조건일 때 쓰면 좋을 것 같습니다.

참석자 3
저희가 지금 추천 로직을 계산할 때 20명의 참가자가 있으면 14,400개의 롤을 조인해야 하는데, 이때는 뭐가 좋을까요?

참석자 1
그때는 Nested Loop Join도 괜찮을 것 같은데요. 20명이...

참석자 3
14,400개의 스케줄에 대해 가능한 날짜를 스케줄 PK로 조인하는 거예요.

참석자 1
양쪽의 크기가 어느 정도 되죠?

참석자 3
1쪽은 4,400개고, 다른 쪽은 아마 14개 정도 될 거예요. 720개라면...

참석자 1
그 개수 기준을 제가 정확히 모르겠지만, 해시 조인은 큰 쪽에 적합하고 Nested Loop는 작은 쪽에 적합하다고 알고 있어요. 중간 기준은 잘 모르니까 경험을 통해 판단하셔야 할 것 같네요.

참석자 3
그렇네요. 20명의...

참석자 1
Hash Join에는 여러 가지 방식이 있는 걸 아시나요? 여러 가지가 뭔지는 잘 모르지만, 대충 디스크를 사용하는지 여부에 따라 나뉘는 것 같더라고요. 해시 테이블을 메모리에 만들면 메모리가 부족할 수 있으니까요.

참석자 3
설마 모자라면 디스크에 저장하나요?

참석자 1
네, 메모리에서만 하는 방식도 있고, 메모리가 부족하면 디스크를 사용하는 방식도 있어요.

참석자 3
그럼 디스크를 어떻게 써요? 그냥 일부는 메모리, 일부는 디스크에 저장하는 건가요?

참석자 1
그 동작까지는 자세히 보진 않았지만, 그런 식인 것 같아요.

참석자 2
파티셔닝한다고도 하고, 해시 값에 따라 파티셔닝한다고도 하고 조인한다고 하네요. 그래서 더 비효율적일 수도 있어요.

참석자 1
디스크를 쓰면...

참석자 3
전문가라면 데이터베이스 질문도 다 대답하겠네요.

참석자 1
저는 잘 모르겠는데, 모른다면 너무 모르겠네요. 아무튼 면접 가나요? 아니요. 예정 사항 있나요?
아니요. 중복 데이터와는 연관이 있나요?
중복 데이터요, 조인 기준으로 말씀하시는 거죠?
그거는 생각을 안 해봤는데, 중복 데이터와 어떻게 연관이 있다는 말씀이죠?
성능이나 데이터가 많을 때, 조인 방식에 대해서 뭐가 더 효율적일까요?
중복 데이터가 많을 때 해시 조인은 해시 테이블에 중복 데이터가 많으면 안 좋을 것 같은데 왜죠?
왜냐하면 해시 충돌이 발생했을 때, 중복 데이터가 없으면 좋지만, 중복 데이터가 많으면 체이닝 방식으로 충돌된 값을 링크드 리스트로 관리해야 하니까요.

참석자 1
충돌된 값들을 다... 보통 링크드 리스트를 통해 값을 찾는 데 시간이 더 걸리죠.
여기까지 할까요?
예, 고맙습니다. 알겠습니다.
